pkg/sql/logictest/tests/local/local_test: TestLogic_select_for_update_read_committed failed

pkg/sql/logictest/tests/local/local_test.TestLogic_select_for_update_read_committed [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13024010?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/13024010?buildTab=artifacts#/) on master @ [161d4f528fac818258358b95c29eafce30cf5233](https://github.com/cockroachdb/cockroach/commits/161d4f528fac818258358b95c29eafce30cf5233):


```
=== RUN   TestLogic_select_for_update_read_committed
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/55d594775d3f532e9011405300e3163d/logTestLogic_select_for_update_read_committed3955704416
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:4354: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11915/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed:320: error while processing
    logic.go:4355: 0 remaining async statements, 3 remaining async queries
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/55d594775d3f532e9011405300e3163d/logTestLogic_select_for_update_read_committed3955704416
--- FAIL: TestLogic_select_for_update_read_committed (2.00s)
=== RUN   TestLogic_select_for_update_read_committed/multi_lock
    logic.go:2613: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11915/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed:256: BEGIN ISOLATION LEVEL READ COMMITTED;
        SELECT * FROM abc WHERE a > 4 FOR UPDATE;
        COMMIT;
        expected:
            5  2  7
            6  1  7
            
        but found (query options: "async,rowsort" -> ignore the following ordering of rows) :
            5  3  7
            6  2  7
            
    logic.go:2184: 
         /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11915/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed:320: too many errors encountered, skipping the rest of the input
[14:40:25] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/11915/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed with config local: 55 tests, 2 failures
    --- FAIL: TestLogic_select_for_update_read_committed/multi_lock (0.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_select_for_update_read_committed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34203

pkg/sql/logictest/tests/local/local_test.TestLogic_select_for_update_read_committed [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/13038117?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/13038117?buildTab=artifacts#/) on master @ [37ad01a3972cb4d34bfc6dfb4b9cfcac360b15dd](https://github.com/cockroachdb/cockroach/commits/37ad01a3972cb4d34bfc6dfb4b9cfcac360b15dd):


```
[07:03:52] 	-- OK;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
SELECT * FROM abc WHERE a > 4 FOR UPDATE SKIP LOCKED;
COMMIT;;
[07:03:52] 	-- OK;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
SELECT * FROM bcd WHERE b < 8 FOR UPDATE SKIP LOCKED;
COMMIT;;
[07:03:52] 	-- OK;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
SELECT * FROM abc WHERE a > 4 FOR UPDATE;
COMMIT;;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
SELECT * FROM bcd WHERE b < 8 FOR UPDATE;
COMMIT;;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
UPDATE abc SET b = b + 1 WHERE a > 4 RETURNING *;
COMMIT;;
[07:03:52] BEGIN ISOLATION LEVEL READ COMMITTED;
UPDATE bcd SET c = c + 1 WHERE b < 8 RETURNING *;
COMMIT;;
[07:03:52] SELECT * FROM abc JOIN bcd ON bcd.b = abc.c WHERE a IN (5, 6) FOR UPDATE OF bcd;
[07:03:52] 	-- OK;
[07:03:52] UPDATE abc SET b = b + 1 WHERE a = 5;
[07:03:52] UPDATE bcd SET c = c + 1 WHERE b = 7;
[07:03:52] SELECT * FROM abc WHERE a > 4;
[07:03:52] 	-- OK;
[07:03:52] SELECT * FROM bcd WHERE b < 8;
[07:03:52] 	-- OK;
[07:03:52] COMMIT;
[07:03:52] 	-- FAIL
    logic.go:2613: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3993/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed:256: BEGIN ISOLATION LEVEL READ COMMITTED;
        SELECT * FROM abc WHERE a > 4 FOR UPDATE;
        COMMIT;
        expected:
            5  2  7
            6  1  7
            
        but found (query options: "async,rowsort" -> ignore the following ordering of rows) :
            5  3  7
            6  2  7
            
[07:03:52] 	-- FAIL
    logic.go:2184: 
         /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3993/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed:320: too many errors encountered, skipping the rest of the input
[07:03:52] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3993/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed with config local: 55 tests, 2 failures

    --- FAIL: TestLogic_select_for_update_read_committed/multi_lock (0.22s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_select_for_update_read_committed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed
+++ b/pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed
@@ -268,22 +268,6 @@ COMMIT;
 6  1  5
 7  2  5
 
-query III async,rowsort q12
-BEGIN ISOLATION LEVEL READ COMMITTED;
-UPDATE abc SET b = b + 1 WHERE a > 4 RETURNING *;
-COMMIT;
-----
-5  3  7
-6  2  7
-
-query III async,rowsort q13
-BEGIN ISOLATION LEVEL READ COMMITTED;
-UPDATE bcd SET c = c + 1 WHERE b < 8 RETURNING *;
-COMMIT;
-----
-6  2  5
-7  3  5
-
 user testuser
 
 query IIIIII rowsort
@@ -318,6 +302,91 @@ user root
 awaitquery q10
 
 awaitquery q11
+
+# Test the same as above, but with UPDATE instead of SELECT FOR UPDATE.
+
+user testuser
+
+statement ok
+BEGIN ISOLATION LEVEL READ COMMITTED
+
+# Try locking the same row twice from a single statement.
+query IIIIII rowsort
+SELECT * FROM bcd JOIN abc ON a = d WHERE bcd.b IN (6, 7) FOR UPDATE OF abc
+----
+6  1  5  5  2  7
+7  2  5  5  2  7
+
+# Try locking the same row twice from the same transaction, with another waiter
+# in between the two.
+
+query III rowsort
+SELECT * FROM bcd WHERE b = 7 FOR UPDATE
+----
+7  2  5
+
+user root
+
+query III rowsort
+BEGIN ISOLATION LEVEL READ COMMITTED;
+SELECT * FROM abc WHERE a > 4 FOR UPDATE SKIP LOCKED;
+COMMIT;
+----
+6  1  7
+
+query III rowsort
+BEGIN ISOLATION LEVEL READ COMMITTED;
+SELECT * FROM bcd WHERE b < 8 FOR UPDATE SKIP LOCKED;
+COMMIT;
+----
+6  1  5
+
+query III async,rowsort q12
+BEGIN ISOLATION LEVEL READ COMMITTED;
+UPDATE abc SET b = b + 1 WHERE a > 4 RETURNING *;
+COMMIT;
+----
+5  4  7
+6  2  7
+
+query III async,rowsort q13
+BEGIN ISOLATION LEVEL READ COMMITTED;
+UPDATE bcd SET c = c + 1 WHERE b < 8 RETURNING *;
+COMMIT;
+----
+6  2  5
+7  4  5
+
+user testuser
+
+query IIIIII rowsort
+SELECT * FROM abc JOIN bcd ON bcd.b = abc.c WHERE a IN (5, 6) FOR UPDATE OF bcd
+----
+5  2  7  7  2  5
+6  1  7  7  2  5
+
+statement ok
+UPDATE abc SET b = b + 1 WHERE a = 5
+
+statement ok
+UPDATE bcd SET c = c + 1 WHERE b = 7
+
+query III rowsort
+SELECT * FROM abc WHERE a > 4
+----
+5  3  7
+6  1  7
+
+query III rowsort
+SELECT * FROM bcd WHERE b < 8
+----
+6  1  5
+7  3  5
+
+statement ok
+COMMIT
+
+user root
 
 awaitquery q12
 

